### ISSUE TITLE
Add a way to specify a custom devicePixelRatio for improved performance

### ISSUE THREAD

**Issue by TimvanScherpenzeel** [uber/luma.gl#643]
#### Background 

If I understand it correctly one can currently use the reported `window.devicePixelRatio` by enabling `useDevicePixels` as implemented in #355 or use the automatic fallback of 1. I would like to request the addition of an option that allows for the use of a custom multiplier (often lower than the reported devicePixelRatio).

I've put together a gist showing the devicePixelRatio's of various devices to get a better idea of what is out there. I often naively assumed that smartphones would use something like a devicePixelRatio of 2 but it turns out many use 3 or 4. The idea would generally be to pass in a lower multiplier, say 1.75 or 2, to trade a bit of visual fidelity in exchange for a much better performance.

https://gist.github.com/TimvanScherpenzeel/a95b431336dfe1e0ec8fd9c5725f84c0

Thank you all for the amazing work so far! I am super impressed by the current state of the project and the defined roadmap.

Kind regards,

Tim van Scherpenzeel

#### To Do List 
  - [ ] Add label and assign to milestone 
  - [ ] Coding
  - [ ] Doc update
  - [ ] Whats new update 
  - [ ] Test

**Comment by 1chandu** [comment]
@TimvanScherpenzeel 

That is a detailed table, thanks !!

Please note `window.devicePixelRatio` also changes when you zoom-in/zoom-out (On mac, use cmd + +/- buttons). So not sure how a fixed devicePixelRatio perform when you are zooming in/out. May be use another scale factor that is optionally multiplied with `window.devicePixelRatio`,  which is [1, 0) range and defaults to `1`? 
CC: @ibgreen

**Comment by ibgreen** [comment]
That's a great table, maybe we could link to it in our docs somewhere...

Just brainstorming around how to allow the user to specify the device pixel ratio.

One idea is to simply overload `useDevicePixels` to allow a number in addition to a boolean value.

But how would the user know a good value to select though? Would they programmatically query a version of the attached table?

We could allow the user to provide a function to `useDevicePixels` - it would send in the current DPR and allow the user to return another, e.g letting the user avoid insane value (like 4x).

```
  useDevicePixels: dpr => Math.max(dpr, 2),
```

It might be useful for that function to have access to the overall width/height of the buffer:

```
  useDevicePixels: (dpr, width, height) =>
    (width * dpr > 2048 || height * dpr > 2048) ? 1 : dpr,
```

One could also send in recent fps to enable dynamic throttling of dpr...

```
  useDevicePixels: (dpr, fps) => fps < 20 ? 1 : dpr,
```

One could also calculate the memory cost of the buffer and supply that, etc.

**Comment by TimvanScherpenzeel** [comment]
Thanks! The table is part of some general research I did a while ago to find out what browser features and WebGL extensions were supported on various devices (I used Browserstack and some devices around the office for these tests). The research including the tools are open source and can be found here: https://github.com/timvanscherpenzeel/detect-features. The raw data can be found here: https://github.com/TimvanScherpenzeel/detect-features/tree/master/data and the `package.json` has a few scripts to generate the earlier mentioned tables.

I wasn't aware of the zooming changing the `devicePixelRatio`, that is really quite interesting and something to consider! For my purposes something as simple as `Math.min(window.devicePixelRatio, 2);` would suffice I think but I can imagine that a smarter solution could enable better ways of determining what `devicePixelRatio` is suitable for what device under various circumstances.
